datafusion/common/src/scalar.rs

```diff
@@ -2433,11 +2455,14 @@ impl ScalarValue {
             ScalarValue::LargeBinary(val) => {
                 eq_array_primitive!(array, index, LargeBinaryArray, val)?
             }
-            ScalarValue::List(arr)
-            | ScalarValue::LargeList(arr)
-            | ScalarValue::FixedSizeList(arr) => {
-                let right = array.slice(index, 1);
-                arr == &right
+            ScalarValue::List(arr) => {
+                Self::eq_array_list(&(arr.to_owned() as ArrayRef), array, index)
+            }
+            ScalarValue::LargeList(arr) => {
+                Self::eq_array_list(&(arr.to_owned() as ArrayRef), array, index)
+            }
+            ScalarValue::FixedSizeList(arr) => {
+                Self::eq_array_list(&(arr.to_owned() as ArrayRef), array, index)
```

Comment:
nit: this could be combined to avoid repeated code. (This is the last comment I'll make about this, but it applies to several more parts of your changeset)

```suggestion
            ScalarValue::List(arr)
            | ScalarValue::LargeList(arr)
            | ScalarValue::FixedSizeList(arr) => {
                Self::eq_array_list(&(arr.to_owned() as ArrayRef), array, index)
```

Reply:
Yeah, I tried this, but the args to `List`, `LargeList`, and `FixedSizeList` are all different, so they can't be put in the same match arm.

Reply:
Ah I see now. They are now distinct data types. Nevermind.